ApplicationStatus

dianarg opened the pull request.

dianarg committed:
Add ApplicationStatus class

- Provides an interface from the application to the controller
  for process id, hint, and progress information.

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Use cache line constant in ApplicationStatus

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Finish up docs for ApplicationStatus; make getters const

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Remove get/set_hash from ApplicationStatus

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

cmcantalupo commented on src/ApplicationStatus.hpp:
```diff
@@ -0,0 +1,150 @@
+            struct m_app_status_s
+            {
+                uint32_t hints;
+                int32_t process; // can be negative
```
not really important, but maybe process should come first, since it is set first?

cmcantalupo commented on src/ApplicationStatus.cpp:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        m_buffer[cpu_idx].total_work = work_units;
+        m_buffer[cpu_idx].completed_work = 0;
```
set completed_work to zero prior to setting total_work units to avoid reporting more complete than total.

cmcantalupo commented on src/ApplicationStatus.cpp:
```diff
@@ -0,0 +1,210 @@
+        GEOPM_DEBUG_ASSERT(m_buffer != nullptr, "m_buffer not set");
+        int total_work = m_buffer[cpu_idx].total_work;
+        if (total_work <= 0) {
+            return NAN;
```
avoid early return.

cmcantalupo commented on src/ApplicationStatus.hpp:
```diff
@@ -0,0 +1,150 @@
+            virtual uint64_t get_hint(int cpu_idx) const = 0;
+            /// @brief Get the current hint bits for every CPU.
+            virtual std::vector<uint64_t> get_hint(void) const = 0;
+            /// @brief Reset the total work units for a single CPU to
```
Providing doxygen for @brief, summary, @param and @return would be nice, but maybe not critical for this PR.

dianarg committed:
Address review comments

- Fixed order of writes for progress
- Removed early return
- Added more doxygen

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add region hash methods to ApplicationStatus

- This reverts commit bb473250da06b010fe1d026fa4f67c3c1226df10.

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Remove ApplicationStatus methods returning vectors

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Rename get_work_progress() to get_progress_cpu()

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Fix signed-unsigned comparison in gtest

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
